executor/index_lookup_join.go

```diff
@@ -468,47 +484,53 @@ func (iw *innerWorker) handleTask(ctx context.Context, task *lookUpJoinTask) err
 	keyBuf := make([]byte, 0, 64)
-	for i := 0; i < task.outerResult.NumRows(); i++ {
-		dLookUpKey, err := iw.constructDatumLookupKey(task, i)
-		if err != nil {
-			return nil, err
-		}
-		if dLookUpKey == nil {
-			// Append null to make looUpKeys the same length as outer Result.
-			task.encodedLookUpKeys.AppendNull(0)
-			continue
-		}
-		keyBuf = keyBuf[:0]
-		keyBuf, err = codec.EncodeKey(iw.ctx.GetSessionVars().StmtCtx, keyBuf, dLookUpKey...)
-		if err != nil {
-			return nil, err
-		}
-		// Store the encoded lookup key in chunk, so we can use it to lookup the matched inners directly.
-		task.encodedLookUpKeys.AppendBytes(0, keyBuf)
-		if iw.hasPrefixCol {
-			for i := range iw.outerCtx.keyCols {
-				// If it's a prefix column. Try to fix it.
-				if iw.colLens[i] != types.UnspecifiedLength {
-					ranger.CutDatumByPrefixLen(&dLookUpKey[i], iw.colLens[i], iw.rowTypes[iw.keyCols[i]])
+	for chkIdx := 0; chkIdx < task.outerResult.NumChunks(); chkIdx++ {
+		chk := task.outerResult.GetChunk(chkIdx)
+		numRows := chk.NumRows()
+		for rowIdx := 0; rowIdx < numRows; rowIdx++ {
+			dLookUpKey, err := iw.constructDatumLookupKey(task, chk, chkIdx, rowIdx)
+			if err != nil {
+				return nil, err
+			}
+			if dLookUpKey == nil {
+				// Append null to make looUpKeys the same length as outer Result.
+				task.encodedLookUpKeys[chkIdx].AppendNull(0)
+				continue
+			}
+			keyBuf = keyBuf[:0]
+			keyBuf, err = codec.EncodeKey(iw.ctx.GetSessionVars().StmtCtx, keyBuf, dLookUpKey...)
+			if err != nil {
+				return nil, err
+			}
+			// Store the encoded lookup key in chunk, so we can use it to lookup the matched inners directly.
+			task.encodedLookUpKeys[chkIdx].AppendBytes(0, keyBuf)
+			if iw.hasPrefixCol {
+				for i := range iw.outerCtx.keyCols {
+					// If it's a prefix column. Try to fix it.
+					if iw.colLens[i] != types.UnspecifiedLength {
+						ranger.CutDatumByPrefixLen(&dLookUpKey[i], iw.colLens[i], iw.rowTypes[iw.keyCols[i]])
+					}
 				}
+				// dLookUpKey is sorted and deduplicated at sortAndDedupLookUpContents.
+				// So we don't need to do it here.
 			}
-			// dLookUpKey is sorted and deduplicated at sortAndDedupLookUpContents.
-			// So we don't need to do it here.
+			lookUpContents = append(lookUpContents, &indexJoinLookUpContent{keys: dLookUpKey, row: chk.GetRow(rowIdx)})
 		}
-		lookUpContents = append(lookUpContents, &indexJoinLookUpContent{keys: dLookUpKey, row: task.outerResult.GetRow(i)})
 	}
 
-	task.memTracker.Consume(task.encodedLookUpKeys.MemoryUsage())
+	for i := range task.encodedLookUpKeys {
+		task.memTracker.Consume(task.encodedLookUpKeys[i].MemoryUsage())
+	}
 	return lookUpContents, nil
 }
 
-func (iw *innerWorker) constructDatumLookupKey(task *lookUpJoinTask, rowIdx int) ([]types.Datum, error) {
-	if task.outerMatch != nil && !task.outerMatch[rowIdx] {
+func (iw *innerWorker) constructDatumLookupKey(task *lookUpJoinTask, chk *chunk.Chunk, chkIdx, rowIdx int) ([]types.Datum, error) {
```

Comment:
What about pass the `chunk.RowPtr` instead of `chk *chunk.Chunk, chkIdx, rowIdx int`.

Reply:
1. `chk` can not be removed
2. I think we do not need to build a new `chunk.RowPtr` since the caller holds the `chkIdx` and `rowIdx` instead of `ptr`.

Reply:
Why can't we remove the `chk`? We can use `task.outerResult.GetRow(chunk.RowPtr)` to get a row.